File: adbpg-memory-cli/node/src/agent-config.js
'use strict';

/**
 * Per-agent configuration management.
 *
 * Each agent has its own JSON file at ~/.adbpg-mem/agents/<agent_id>.json
 * with mode 0600. Schema:
 *   - isolation_agent: boolean (default: false)
 *   - isolation_run_mode: enum ["off", "manual", "auto", "tag"] (default: "off")
 *   - current_run_id: string (optional, no default)
 *
 * Defaults are applied at read time when a key is absent. They are NOT
 * persisted to disk on first set — only explicitly-set keys are written.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const AGENTS_DIR = path.join(os.homedir(), '.adbpg-mem', 'agents');

const ISOLATION_RUN_MODES = ['off', 'manual', 'auto', 'tag'];

// Field schema: type + default (default of `undefined` means "no default — omit from show")
const SCHEMA = {
  isolation_agent: {
    type: 'boolean',
    default: false,
  },
  isolation_run_mode: {
    type: 'enum',
    values: ISOLATION_RUN_MODES,
    default: 'off',
  },
  current_run_id: {
    type: 'string',
    // No default; absent from show output when unset.
  },
};

const KNOWN_KEYS = Object.keys(SCHEMA);

// agent_id: 1-64 chars, [a-zA-Z0-9_-]
const AGENT_ID_RE = /^[a-zA-Z0-9_-]{1,64}$/;

/**
 * Validate the agent_id format. Returns null if valid, else an error message.
 * @param {*} agentId
 * @returns {string|null}
 */
function validateAgentId(agentId) {
  if (agentId === undefined || agentId === null || agentId === '') {
    return 'agent_id is required';
  }
  if (typeof agentId !== 'string') {
    return 'agent_id must be a string';
  }
  if (agentId.length > 64) {
    return 'agent_id exceeds 64 characters';
  }
  if (!AGENT_ID_RE.test(agentId)) {
    return 'agent_id must match [a-zA-Z0-9_-] and be 1-64 chars';
  }
  return null;
}

/**
 * Resolve the on-disk path for a given agent_id under a base directory.
 * @param {string} agentId
 * @param {string} [baseDir] - Defaults to AGENTS_DIR.
 * @returns {string}
 */
function agentConfigPath(agentId, baseDir) {
  const dir = baseDir != null ? baseDir : AGENTS_DIR;
  return path.join(dir, `${agentId}.json`);
}

/**
 * Load raw stored config for an agent. Returns {} if file missing.
 * @param {string} agentId
 * @param {string} [baseDir]
 * @returns {object}
 */
function loadAgentConfig(agentId, baseDir) {
  const filePath = agentConfigPath(agentId, baseDir);
  if (!fs.existsSync(filePath)) {
    return {};
  }
  const raw = fs.readFileSync(filePath, 'utf-8');
  return JSON.parse(raw);
}

/**
 * Persist agent config to disk with 0600 perms. Creates the agents dir
 * (with default perms) on demand. Atomically rewrites the file each time.
 * @param {string} agentId
 * @param {object} config
 * @param {string} [baseDir]
 */
function saveAgentConfig(agentId, config, baseDir) {
  const filePath = agentConfigPath(agentId, baseDir);
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(config, null, 2), { encoding: 'utf-8', mode: 0o600 });
  // Ensure perms even when file pre-existed with looser mode.
  try {
    fs.chmodSync(filePath, 0o600);
  } catch (_) {
    // best-effort
  }
}

/**
 * Remove an agent's config file. No-op if missing.
 * @param {string} agentId
 * @param {string} [baseDir]
 */
function removeAgentConfig(agentId, baseDir) {
  const filePath = agentConfigPath(agentId, baseDir);
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
}

/**
 * Build the effective view of an agent's config: stored values overlaid on
 * schema defaults. Fields without a default and without a stored value
 * are omitted (not surfaced as null/undefined).
 * @param {object} stored
 * @returns {object}
 */
function applyDefaults(stored) {
  const out = {};
  for (const [key, def] of Object.entries(SCHEMA)) {
    if (Object.prototype.hasOwnProperty.call(stored, key)) {
      out[key] = stored[key];
    } else if (def.default !== undefined) {
      out[key] = def.default;
    }
  }
  return out;
}

/**
 * Parse and validate a CLI-supplied string value for a known key.
 * Returns { ok: true, value } or { ok: false, error }.
 *
 * Boolean parsing accepts: true/false (case-insensitive). Anything else errors.
 * Enum parsing requires exact match against the allowed set.
 * String parsing passes the value through unchanged (length>0 enforced for
 * current_run_id since empty would be indistinguishable from "unset" intent).
 *
 * @param {string} key
 * @param {string} rawValue
 * @returns {{ ok: boolean, value?: any, error?: string }}
 */
function parseValue(key, rawValue) {
  if (!Object.prototype.hasOwnProperty.call(SCHEMA, key)) {
    return { ok: false, error: `unknown key '${key}'. Known keys: ${KNOWN_KEYS.join(', ')}` };
  }
  const def = SCHEMA[key];

  if (def.type === 'boolean') {
    if (typeof rawValue !== 'string') {
      return { ok: false, error: `invalid value for ${key}: expected boolean string, got ${typeof rawValue}` };
    }
    const lowered = rawValue.toLowerCase();
    if (lowered === 'true') return { ok: true, value: true };
    if (lowered === 'false') return { ok: true, value: false };
    return { ok: false, error: `invalid value for ${key}: expected 'true' or 'false', got '${rawValue}'` };
  }

  if (def.type === 'enum') {
    if (!def.values.includes(rawValue)) {
      return {
        ok: false,
        error: `invalid value for ${key}: must be one of ${def.values.join(', ')}, got '${rawValue}'`,
      };
    }
    return { ok: true, value: rawValue };
  }

  if (def.type === 'string') {
    if (typeof rawValue !== 'string' || rawValue === '') {
      return { ok: false, error: `invalid value for ${key}: expected non-empty string` };
    }
    return { ok: true, value: rawValue };
  }

  // Defensive — schema only defines the three types above.
  return { ok: false, error: `invalid value for ${key}: unsupported type` };
}

/**
 * High-level operations used by the CLI layer. Each returns
 * { ok: true, data } or { ok: false, error }. The caller is responsible
 * for shaping the JSON envelope / text output.
 */

function opShow(agentId, baseDir) {
  const idErr = validateAgentId(agentId);
  if (idErr) return { ok: false, error: invalidIdMsg(idErr) };
  const stored = loadAgentConfig(agentId, baseDir);
  return { ok: true, data: applyDefaults(stored) };
}

function opGet(agentId, key, baseDir) {
  const idErr = validateAgentId(agentId);
  if (idErr) return { ok: false, error: invalidIdMsg(idErr) };
  if (!Object.prototype.hasOwnProperty.call(SCHEMA, key)) {
    return { ok: false, error: `unknown key '${key}'. Known keys: ${KNOWN_KEYS.join(', ')}` };
  }
  const view = applyDefaults(loadAgentConfig(agentId, baseDir));
  // current_run_id may legitimately be absent — surface as null in the data shape.
  const value = Object.prototype.hasOwnProperty.call(view, key) ? view[key] : null;
  return { ok: true, data: { key, value } };
}

function opSet(agentId, key, rawValue, baseDir) {
  const idErr = validateAgentId(agentId);
  if (idErr) return { ok: false, error: invalidIdMsg(idErr) };
  const parsed = parseValue(key, rawValue);
  if (!parsed.ok) return { ok: false, error: parsed.error };
  const stored = loadAgentConfig(agentId, baseDir);
  stored[key] = parsed.value;
  saveAgentConfig(agentId, stored, baseDir);
  return { ok: true, data: { key, value: parsed.value } };
}

function opUnset(agentId, key, baseDir) {
  const idErr = validateAgentId(agentId);
  if (idErr) return { ok: false, error: invalidIdMsg(idErr) };
  if (!Object.prototype.hasOwnProperty.call(SCHEMA, key)) {
    return { ok: false, error: `unknown key '${key}'. Known keys: ${KNOWN_KEYS.join(', ')}` };
  }
  const filePath = agentConfigPath(agentId, baseDir);
  if (!fs.existsSync(filePath)) {
    // Idempotent: nothing to unset.
    return { ok: true, data: { key, removed: false } };
  }
  const stored = loadAgentConfig(agentId, baseDir);
  let removed = false;
  if (Object.prototype.hasOwnProperty.call(stored, key)) {
    delete stored[key];
    removed = true;
  }
  saveAgentConfig(agentId, stored, baseDir);
  return { ok: true, data: { key, removed } };
}

function invalidIdMsg(detail) {
  return `invalid agent_id format: ${detail}`;
}

module.exports = {
  AGENTS_DIR,
  ISOLATION_RUN_MODES,
  SCHEMA,
  KNOWN_KEYS,
  AGENT_ID_RE,
  validateAgentId,
  agentConfigPath,
  loadAgentConfig,
  saveAgentConfig,
  removeAgentConfig,
  applyDefaults,
  parseValue,
  opShow,
  opGet,
  opSet,
  opUnset,
};

File: adbpg-memory-cli/python/src/adbpg_memory_cli/agent_config.py
"""Per-agent configuration for ADBPG Memory CLI.

Stores agent-specific settings under ``~/.adbpg-mem/agents/<agent_id>.json``
with file mode ``0600``. Schema:

    isolation_agent:    bool, default False
    isolation_run_mode: enum {"off","manual","auto","tag"}, default "off"
    current_run_id:     str, optional (omitted from defaults)

Unknown keys or values that do not satisfy the schema raise
``AgentConfigError`` so the CLI can surface ``status=error``.
"""

import json
import os
import re
from pathlib import Path
from typing import Any

from .config import CONFIG_DIR

AGENTS_DIR = CONFIG_DIR / "agents"

# ---------------------------------------------------------------------------
# Schema
# ---------------------------------------------------------------------------

VALID_KEYS = ("isolation_agent", "isolation_run_mode", "current_run_id")
VALID_RUN_MODES = ("off", "manual", "auto", "tag")

DEFAULTS: dict[str, Any] = {
    "isolation_agent": False,
    "isolation_run_mode": "off",
}

_AGENT_ID_RE = re.compile(r"^[A-Za-z0-9_-]+$")


class AgentConfigError(ValueError):
    """Raised for any schema/validation failure in agent-config commands."""


# ---------------------------------------------------------------------------
# Validation helpers
# ---------------------------------------------------------------------------

def validate_agent_id(agent_id: str | None) -> str:
    """Return ``agent_id`` if valid, otherwise raise ``AgentConfigError``.

    Empty/None is treated as missing rather than an invalid format so callers
    can surface the user-facing "agent-config commands require -a <agent_id>"
    error separately.
    """
    if agent_id is None or agent_id == "":
        raise AgentConfigError("agent-config commands require -a <agent_id>")
    if len(agent_id) > 64:
        raise AgentConfigError(
            f"invalid agent_id format: length {len(agent_id)} exceeds max 64"
        )
    if not _AGENT_ID_RE.match(agent_id):
        raise AgentConfigError(
            "invalid agent_id format: only [A-Za-z0-9_-] characters are allowed"
        )
    return agent_id


def _coerce_value(key: str, value: Any) -> Any:
    """Coerce a CLI-provided string value into the schema-typed value.

    Non-string inputs (e.g. already-typed values from tests) are validated
    against the schema as-is. Raises ``AgentConfigError`` on type mismatch.
    """
    if key == "isolation_agent":
        # Native bools (used by the Python API / tests) are accepted as-is.
        # Strings must be exactly "true"/"false" (case-insensitive) — values
        # like "1"/"0"/"yes"/"no" are rejected to keep parity with the Node
        # implementation (see adbpg-memory-cli/node/src/agent-config.js: the
        # `bool` parser only accepts true/false). The error message text is
        # an explicit cross-impl contract.
        if isinstance(value, bool):
            return value
        if isinstance(value, str):
            lowered = value.lower()
            if lowered == "true":
                return True
            if lowered == "false":
                return False
            raise AgentConfigError(
                f"invalid value for isolation_agent: expected 'true' or 'false', got '{value}'"
            )
        raise AgentConfigError(
            f"invalid value for isolation_agent: expected 'true' or 'false', got '{value}'"
        )

    if key == "isolation_run_mode":
        if not isinstance(value, str):
            raise AgentConfigError(
                f"invalid value for isolation_run_mode: expected string, got {type(value).__name__}"
            )
        if value not in VALID_RUN_MODES:
            raise AgentConfigError(
                f"invalid value for isolation_run_mode: must be one of "
                f"{list(VALID_RUN_MODES)}, got {value!r}"
            )
        return value

    if key == "current_run_id":
        if not isinstance(value, str):
            raise AgentConfigError(
                f"invalid value for current_run_id: expected string, got {type(value).__name__}"
            )
        if value == "":
            raise AgentConfigError(
                "invalid value for current_run_id: must be non-empty"
            )
        return value

    raise AgentConfigError(f"unknown key: {key!r}")


def validate_key(key: str) -> str:
    if key not in VALID_KEYS:
        raise AgentConfigError(f"unknown key: {key!r}")
    return key


# ---------------------------------------------------------------------------
# Persistence
# ---------------------------------------------------------------------------

def _agent_file(agent_id: str, agents_dir: Path | None = None) -> Path:
    base = agents_dir if agents_dir is not None else AGENTS_DIR
    return base / f"{agent_id}.json"


def _load_raw(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def _save_raw(path: Path, data: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    # Use os.open with mode 0600 so the file is created with restrictive
    # permissions on POSIX. On Windows the mode argument is ignored.
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
    except Exception:
        # Best-effort cleanup if fdopen raised after os.open succeeded.
        try:
            os.close(fd)
        except OSError:
            pass
        raise
    # Defensive: ensure mode is 0600 even if file pre-existed with looser mode.
    try:
        os.chmod(path, 0o600)
    except OSError:
        pass


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def show_agent_config(
    agent_id: str, agents_dir: Path | None = None
) -> dict[str, Any]:
    """Return the merged config for ``agent_id`` (defaults + stored overrides).

    Missing file → defaults only (current_run_id absent).
    """
    validate_agent_id(agent_id)
    stored = _load_raw(_agent_file(agent_id, agents_dir))
    merged: dict[str, Any] = dict(DEFAULTS)
    for k in VALID_KEYS:
        if k in stored:
            merged[k] = stored[k]
    return merged


def get_agent_config(
    agent_id: str, key: str, agents_dir: Path | None = None
) -> Any:
    """Return a single key value for ``agent_id``.

    Falls back to the schema default. If the key has no default and is unset,
    returns ``None``.
    """
    validate_agent_id(agent_id)
    validate_key(key)
    stored = _load_raw(_agent_file(agent_id, agents_dir))
    if key in stored:
        return stored[key]
    return DEFAULTS.get(key)


def set_agent_config(
    agent_id: str, key: str, value: Any, agents_dir: Path | None = None
) -> Any:
    """Persist ``key=value`` for ``agent_id``. Returns the coerced value."""
    validate_agent_id(agent_id)
    validate_key(key)
    coerced = _coerce_value(key, value)
    path = _agent_file(agent_id, agents_dir)
    stored = _load_raw(path)
    stored[key] = coerced
    _save_raw(path, stored)
    return coerced


def unset_agent_config(
    agent_id: str, key: str, agents_dir: Path | None = None
) -> None:
    """Delete ``key`` from the agent config (idempotent)."""
    validate_agent_id(agent_id)
    validate_key(key)
    path = _agent_file(agent_id, agents_dir)
    if not path.exists():
        return
    stored = _load_raw(path)
    if key in stored:
        del stored[key]
        _save_raw(path, stored)
